numpy/core/src/umath/ufunc_object.c

```diff
@@ -646,253 +726,206 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
     }
 
     /*
-     * Get keyword output and other arguments.
-     * Raise an error if anything else is present in the
-     * keyword dictionary.
+     * If keywords are present, get keyword output and other arguments.
+     * Raise an error if anything else is present in the keyword dictionary.
      */
-    if (kwds != NULL) {
+    if (kwds == NULL) {
+        return 0; /* no keywords, we're done */
+    }
+    else {
         PyObject *key, *value;
         Py_ssize_t pos = 0;
-        while (PyDict_Next(kwds, &pos, &key, &value)) {
-            Py_ssize_t length = 0;
-            char *str = NULL;
-            int bad_arg = 1;
+        PyObject *out_kwd = NULL;
+        PyArray_Descr *dtype = NULL;
+        PyObject *sig = NULL;
+        void *outputs[13];
+        /*
+         * We store the outputs in order, so we can easily set them
+         * in our loop.  Note that if any of the out_* is NULL,
+         * the keyword will be treated as unwanted.
+         */
+        outputs[0] = &out_kwd;
+        outputs[1] = out_wheremask;
+        outputs[2] = out_axes;
+        outputs[3] = out_axis;
+        outputs[4] = out_keepdims;
+        outputs[5] = out_casting;
+        outputs[6] = out_order;
+        outputs[7] = &dtype;
+        outputs[8] = out_subok;
+        outputs[9] = out_typetup;
+        outputs[10] = &sig;
+        outputs[11] = out_extobj;
+        outputs[12] = NULL;  /* sentinel for key-not-found */
 
-#if defined(NPY_PY3K)
-            Py_XDECREF(str_key_obj);
-            str_key_obj = PyUnicode_AsASCIIString(key);
-            if (str_key_obj != NULL) {
-                key = str_key_obj;
+        while (PyDict_Next(kwds, &pos, &key, &value)) {
+            /*
+             * For maximum speed, we follow CPython's ceval.c
+             * _PyEval_EvalCodeWithName, which tries direct pointer
+             * comparison first, noting:
+             * """Speed hack: do raw pointer compares. As names are
+             *    normally interned this should almost always hit."""
+             * And we do indeed intern them.
+             */
+            void *output;
+            ufunc_kwarg *kwarg = ufunc_kwargs;
```

Comment:
Can you extract this line to line 787 as a helper function, `ufunc_kwarg *find_ufunc_kwd(PyObject *key)`, which returns null if nothing is found?

That would make it easier to swap it for a pointer-ordered binary search in future, if necessary